### PR TITLE
Add Fabric implementation for Android

### PR DESCRIPTION
**Pull request by zolbooo**
Hello, thank you for work on this project 👋

I have updated example app to use React Native `0.69.0-rc.6`, which is the first release to enable new architecture by default. Also I have added implementation of VectorDrawable component for Fabric. I have build application with and without new arch enabled and example app seems to work correctly.

Test plan:
- [x] Android - Fabric example app
- [x] Android - Paper example app (built with `newArchEnabled=false` in `gradle.properties`
- [x] iOS - Fabric app (`USE_FABRIC=1 RCT_NEW_ARCH_ENABLED=1 pod install`)
- [x] iOS - Paper app (Regular `pod install`)
- [ ] Some props could be implemented incorrectly because `RNVectorDrawableNativeComponent` and `index.d.ts` typings differ

### CODE REVIEW

**Comment by zolbooo** [comment]
Could you review this, @oblador?

File: android/build.gradle [hunk 1]
```diff
@@ -23,6 +30,46 @@ android {
...
+        if (isNewArchitectureEnabled()) {
+            var appProject = rootProject.allprojects.find { it.plugins.hasPlugin('com.android.application') }
+            externalNativeBuild {
+                ndkBuild {
+                    arguments "APP_PLATFORM=android-21",
+                            "APP_STL=c++_shared",
+                            "NDK_TOOLCHAIN_VERSION=clang",
+                            "GENERATED_SRC_DIR=${appProject.buildDir}/generated/source",
+                            "PROJECT_BUILD_DIR=${appProject.buildDir}",
+                            "REACT_ANDROID_DIR=${appProject.rootDir}/../node_modules/react-native/ReactAndroid",
```
Comment (oblador) [inline android/build.gradle]: This is an unsafe assumption about where react-native package is located that works with this project, but breaks mono-repos (that we use at Klarna). See this [gradle file](https://github.com/oblador/react-native-performance/blob/master/packages/react-native-performance/android/build.gradle#L15-L28) for an example on how that can be done in a more dynamic fashion. I've discussed this with Meta but no better solution in place than this from what I know.

File: android/build.gradle [hunk 2]
```diff
@@ -23,6 +30,46 @@ android {
...
+
+    if (isNewArchitectureEnabled()) {
+        externalNativeBuild {
+            ndkBuild {
+                path "src/main/jni/Android.mk"
+            }
+        }
+    }
+    packagingOptions {
+        // See: https://github.com/software-mansion/react-native-screens/blob/main/android/build.gradle#L90
```
Comment (oblador) [inline android/build.gradle:94]: Weird, but ok

File: android/build.gradle [hunk 3]
```diff
@@ -32,18 +79,24 @@ android {
 }
 
 repositories {
-    if (project == rootProject) {
```
Comment (oblador) [inline android/build.gradle]: Any particular reason why you changed this setup?

File: android/build.gradle [hunk 4]
```diff
@@ -32,18 +79,24 @@ android {
...
 }
 
 dependencies {
     //noinspection GradleDynamicVersion
     implementation "com.facebook.react:react-native:+"
     implementation 'androidx.appcompat:appcompat:1.2.0'
 }
+
+if (isNewArchitectureEnabled()) {
+    react {
```
Comment (oblador) [inline android/build.gradle:131]: For monorepo setups I believe you need to pass reactNativeDir/codegenDir too as in the linked example

File: android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableManager.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import com.facebook.react.uimanager.ThemedReactContext;
+import com.facebook.react.uimanager.ViewProps;
+import com.facebook.react.uimanager.ViewManagerDelegate;
+import com.facebook.react.uimanager.annotations.ReactProp;
+import com.facebook.react.viewmanagers.RNVectorDrawableManagerDelegate;
+import com.facebook.react.viewmanagers.RNVectorDrawableManagerInterface;
+
+@ReactModule(name = VectorDrawableManagerImpl.NAME)
+public class VectorDrawableManager extends SimpleViewManager<ImageView>
+		implements RNVectorDrawableManagerInterface<ImageView> {
```
Comment (oblador) [inline android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableManager.java]: Mixing of spaces and tab indentation in this file, please make sure all java files are using 4 spaces

File: android/src/main/jni/VectorDrawableComponentsRegistry.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    return registry;
+  };
+
+  delegate->buildRegistryFunction = buildRegistryFunction;
+  return instance;
+}
+
+void VectorDrawableComponentsRegistry::registerNatives() {
+  registerHybrid({
+      makeNativeMethod("initHybrid", VectorDrawableComponentsRegistry::initHybrid),
```
Comment (oblador) [inline android/src/main/jni/VectorDrawableComponentsRegistry.cpp:47]:
```suggestion
    makeNativeMethod("initHybrid", VectorDrawableComponentsRegistry::initHybrid),
```

File: android/src/main/jni/VectorDrawableComponentsRegistry.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  const ComponentFactory *delegate_;
+
+  static jni::local_ref<jhybriddata> initHybrid(
+      jni::alias_ref<jclass>,
+      ComponentFactory *delegate);
+};
+
+} // namespace react
+} // namespace facebook
```
Comment (oblador) [inline android/src/main/jni/VectorDrawableComponentsRegistry.h:34]:
```suggestion
} // namespace facebook

```

File: android/src/main/jni/VectorDrawableComponentsRegistry.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  });
+  // See: https://github.com/software-mansion/react-native-screens/blob/main/android/src/main/jni/RNScreensComponentsRegistry.cpp#L62
+  sharedProviderRegistry();
+}
+
+} // namespace react
+} // namespace facebook
```
Comment (oblador) [inline android/src/main/jni/VectorDrawableComponentsRegistry.cpp:54]:
```suggestion
} // namespace facebook

```

File: android/src/main/jni/OnInit.cpp [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#include <fbjni/fbjni.h>
+
+#include "VectorDrawableComponentsRegistry.h"
+
+JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *) {
+  return facebook::jni::initialize(vm, [] {
+    facebook::react::VectorDrawableComponentsRegistry::registerNatives();
+  });
+}
```
Comment (oblador) [inline android/src/main/jni/OnInit.cpp]:
```suggestion
}

```

File: android/src/main/jni/Android.mk [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    libreact_render_componentregistry \
+    libreact_render_debug \
+    libruntimeexecutor \
+    libreact_render_mapbuffer \
+    libreact_codegen_rncore \
+    libreact_codegen_vectordrawable
+
+LOCAL_CFLAGS := -DLOG_TAG=\"ReactNative\" -fexceptions -frtti -std=c++17 -Wall
+
+include $(BUILD_SHARED_LIBRARY)
```
Comment (oblador) [inline android/src/main/jni/Android.mk]:
```suggestion
include $(BUILD_SHARED_LIBRARY)

```

File: android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableComponentsRegistry.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  @DoNotStrip
+  private VectorDrawableComponentsRegistry(ComponentFactory componentFactory) {
+    mHybridData = initHybrid(componentFactory);
+  }
+
+  @DoNotStrip
+  public static VectorDrawableComponentsRegistry register(ComponentFactory componentFactory) {
+    return new VectorDrawableComponentsRegistry(componentFactory);
+  }
+}
```
Comment (oblador) [inline android/src/fabric/java/com/klarna/vectordrawable/VectorDrawableComponentsRegistry.java]: For the pedantic smurf inside me, please always have an empty new line at the end of the file 🙏 
```suggestion
}

```

File: example/package.json [hunk 12]
```diff
@@ -11,9 +11,11 @@
     "lint": "eslint ."
   },
   "dependencies": {
-    "react": "17.0.1",
-    "react-native": "0.64.0",
-    "@klarna/react-native-vector-drawable": "*"
+    "@klarna/react-native-vector-drawable": "*",
+    "react": "18.0.0",
+    "react-native": "0.69.0-rc.6",
+    "react-native-codegen": "^0.69.1",
```
Comment (oblador) [inline example/package.json]: I didn't think it was necessary to explicitly include codegen/gradle-plugin dependencies

File: src/RNVectorDrawableNativeComponent.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import {
+  ViewProps,
+  FlexStyle,
+  TransformsStyle,
+  ColorValue,
+  HostComponent,
+} from 'react-native';
+
+// @ts-ignore
+import {WithDefault} from 'react-native/Libraries/Types/CodegenTypes';
```
Comment (oblador) [inline src/RNVectorDrawableNativeComponent.ts]:
```suggestion
import { WithDefault } from 'react-native/Libraries/Types/CodegenTypes';
```

**Review by oblador — CHANGES_REQUESTED** [review]
Absolutely fantastic PR! Added a bunch of comments, mostly minor, but I'm missing updated README instructions and maybe a mention of Fabric support.

**Comment by oblador** [comment]
@zolbooo I'll be going on vacation tomorrow just as a heads up. 👍

**Comment by zolbooo** [comment]
Okay, I will push my commits in a few hours.

**Comment by zolbooo** [comment]
Everything seems good to me now, please check 🚀

**Comment by oblador** [comment]
Back from vacation now :D I tried to run the example project with fabric turned off (hence using paper), but it failed to compile with this error:
```
node_modules/@klarna/react-native-vector-drawable/android/src/paper/java/com/klarna/vectordrawable/VectorDrawableManager.java:22: error: invalid method declaration; return type required
    public ColoredViewManager(ReactApplicationContext reactContext) {
           ^
1 error
```

I didn't look much further into this error, but I'm guessing it's because codegen not being run in old architecture.

**Comment by zolbooo** [comment]
> Back from vacation now :D I tried to run the example project with fabric turned off (hence using paper), but it failed to compile with this error:
> 
> ```
> 
> node_modules/@klarna/react-native-vector-drawable/android/src/paper/java/com/klarna/vectordrawable/VectorDrawableManager.java:22: error: invalid method declaration; return type required
> 
>     public ColoredViewManager(ReactApplicationContext reactContext) {
> 
>            ^
> 
> 1 error
> 
> ```
> 
> 
> 
> I didn't look much further into this error, but I'm guessing it's because codegen not being run in old architecture. 

Oops, I forgot to rename constructor in Paper implementation. Please check the latest commit.

**Comment by oblador** [comment]
Still same error for me 🤔 Did you manage to run this locally?

**Comment by zolbooo** [comment]
My bad, I was editing `node_modules` in example instead of editing library files directly. Now it compiles on both archtitectures. Also added support for M1 devices in [4b47a91](https://github.com/klarna-incubator/react-native-vector-drawable/pull/5/commits/4b47a91d9df47057fb36c954323f60a7bb1c6e64).

**Review by oblador — APPROVED** [review]

**Pull request merged**